### PR TITLE
feat: improve stdlib base64 performance by 98%+

### PR DESCRIPTION
**Pull request by ghostsquad**
```
benchmark                                 old ns/op       new ns/op     delta
Benchmark_Builtin_base64-16               10805730974     23158636      -99.79%
Benchmark_Builtin_base64_bytearray-16     8682808704      123360964     -98.58%
```

### Notes

~This PR contains commits from another pending PR. I'll rebased once #343 is merged.~

### CODE REVIEW

**Comment by ghostsquad** [comment]
resolves #341

File: builtins.go [hunk 1]
```diff
@@ -664,6 +665,16 @@ func builtinMd5(i *interpreter, trace traceElement, x value) (value, error) {
 	return makeValueString(hex.EncodeToString(hash[:])), nil
 }
 
+func builtinBase64(i *interpreter, trace traceElement, input value) (value, error) {
+	str, err := i.getString(input, trace)
```
Comment (sbarzowski) [inline builtins.go]: This should also work with arrays  of bytes, see: https://github.com/google/jsonnet/blob/552d8ec6f6b973a6357b83eb9bacd707366d28f0/stdlib/std.jsonnet#L1072

The string version should error out when there are characters with codepoints > 255 and those <= should use codepoints-as-bytes encoding (the current implementation results in UTF8 encoding, which would be a better option, if we didn't need to worry about backwards compatibility). See: https://github.com/google/jsonnet/blob/552d8ec6f6b973a6357b83eb9bacd707366d28f0/stdlib/std.jsonnet#L1108

And we should have tests to cover all that :-).

Comment (ghostsquad) [reply builtins.go]: Thanks for the feedback. I'll add that functionality and add tests as well.

Comment (ghostsquad) [reply builtins.go]: done

Comment (ghostsquad) [reply builtins.go]: implemented for arrays and strings

**Comment by ghostsquad** [comment]
@sbarzowski this has been rebased and is ready for review/merge

File: builtins.go [hunk 2]
```diff
@@ -664,6 +665,63 @@ func builtinMd5(i *interpreter, trace traceElement, x value) (value, error) {
...
+		if !(v < 256 && v >= 0) {
+			msg := fmt.Sprintf("base64 encountered invalid codepoint value in the array (must be 0 <= X <= 255), got %d", v)
+			return msg, false
+		}
+
+		return "", true
+	}
+
+	vStr, err := i.getString(input, trace)
+	if err == nil {
```
Comment (sbarzowski) [inline builtins.go]: I don't like the assumption that any error returned by `i.getString` means simply that input was not a string. This function returns a generic error which may be anything in principle. If another error condition is currently possible or will be possible in the future versions, we'll have a really confusing bug here.

One more elegant approach would be to use a type switch here. Example: https://github.com/google/go-jsonnet/blob/v0.14.0/builtins.go#L44.

File: builtins.go [hunk 3]
```diff
@@ -664,6 +665,63 @@ func builtinMd5(i *interpreter, trace traceElement, x value) (value, error) {
 	return makeValueString(hex.EncodeToString(hash[:])), nil
 }
 
+func builtinBase64(i *interpreter, trace traceElement, input value) (value, error) {
+	var byteArr []byte
+
+	var sanityCheck = func(v int) (string, bool) {
+		if !(v < 256 && v >= 0) {
```
Comment (sbarzowski) [inline builtins.go]: A very minor nit: I think a more idiomatic way of checking if value is in the range would be:
`0 <= v && v < 256`. It's the closest you can get in most programming languages to the mathy notation 0 <= v < 256. 

Alternatively eliminating the negation would also make it more clear: `v < 0 || v > 256` than the current expresssion.

File: builtins.go (hunk 2 above)
Comment (ghostsquad) [reply builtins.go]: that looks good. I'll update

File: builtins.go (hunk 3 above)
Comment (ghostsquad) [reply builtins.go]: I like the `||` variant, might be written as `0 > v || v > 256`

Comment (ghostsquad) [reply builtins.go]: no.. that is a little confusing to look at. I'm going to use `v < 0 || v > 255`

File: builtins.go [hunk 4]
```diff
@@ -681,6 +682,69 @@ func builtinMd5(i *interpreter, trace traceElement, x value) (value, error) {
...
+		if v < 0 || v > 256 {
+			msg := fmt.Sprintf("base64 encountered invalid codepoint value in the array (must be 0 <= X <= 255), got %d", v)
+			return msg, false
+		}
+
+		return "", true
+	}
+
+	switch inputType := input.(type) {
+	case *valueString:
```
Comment (sbarzowski) [inline builtins.go]: valueString is an interface type these days, so no * here.

Comment (ghostsquad) [reply builtins.go]: fixed

File: builtins.go [hunk 5]
```diff
@@ -681,6 +682,69 @@ func builtinMd5(i *interpreter, trace traceElement, x value) (value, error) {
 	return makeValueString(hex.EncodeToString(hash[:])), nil
 }
 
+func builtinBase64(i *interpreter, trace traceElement, input value) (value, error) {
+	var byteArr []byte
+
+	var sanityCheck = func(v int) (string, bool) {
+		if v < 0 || v > 256 {
```
Comment (cbandy) [inline builtins.go]: This was originally `!(v < 256 && v >= 0)`. If I understand correctly, the same logic should be `v >= 256` or `v > 255` here.

One more bike shed... 😝 I prefer the mathy/range look of less-than: `v < 0 || 255 < v`

Comment (ghostsquad) [reply builtins.go]: oh yes.. the number should be 255.

Comment (ghostsquad) [reply builtins.go]: I agree that `N < v && v < M` looks best. However, it reads differently when it's an `||` (or). And I think `v < N || v > M` feels better.

Comment (ghostsquad) [reply builtins.go]: wait.. I now see the desired range in `v < 0 || 255 < v`...

**Pull request merged**